Annotation parallelization

nictru opened the pull request:
The annotation currently works by going through a - potentially very large - file containing circRNA location and investigating each of them separately. This changes introduce parallelization to this process

nictru committed:
Implement multi-threading in annotation

nictru committed:
Switch to annotating only combined circRNAs

nictru committed:
Parallelize annotation chromosome-wise

nictru committed:
Fix annotation channel structure

rreggiar commented on modules/local/annotation/full_annotation/main.nf:
```diff
@@ -30,7 +27,12 @@ process ANNOTATION {
-    annotate_outputs.sh $exon_boundary &> ${prefix}.log
+    mkdir -p bed12
+
+    parallel -j $task.cpus -a circs.bed annotate_outputs.sh $exon_boundary {}
```
Since order doesn't matter, you could use `parallel -u` to speed up output. I went with a split > parallel > pool approach in [my fork](https://github.com/rreggiar/circrna/commit/904104a647fc82fdf93724f452231fe2df9cb710#diff-f53b1433313394a305660960d73aec029377fdec571a89a3359e7b4a4fb14fa6R41-R48) and feel like the implementation in this commit can be optimized, but as long as this works for now we can worry about that later.

rreggiar reviewed (approved):
IMO this is critical for more complex data, glad we both arrived at the solution separately. Everything makes sense -- I left some notes on the parallel implementation but that is an enhancement we can think about later as long as this works.

nictru commented:
Answer to the comments from #85

I switched to only annotating the combined circRNAs because previously this was done for each detection tool for each sample. This led to a large number of long-running tasks, which most likely had a large overlap. I understand that it might be interesting to have tool-specific annotations, but I think this should then be approached like this:

1. Union of all circRNAs
2. Split into multiple files (by chromosome or line number)
3. Annotate
4. Union of annotations
5. For each sample-tool combination extract the fitting annotations

The core problem here is, that even with the parallelized annotation, the annotation can take hours per task. 

Notes on a potentially more elegant approach:
I tried using `bedtools intersect` directly on all circRNAs and the GTF file, which was super fast in comparison. The only problem here is, that we need to do some aggregation of all the hits for each circRNA afterwards. I can imagine that this could work well using `pandas.groupBy`.

rreggiar commented:
Yeah the annotation bottleneck is a major problem right now. Even if we intersect everything together, we'll need to apply the same logic to each group, correct? There may be a more efficient implementation of this, I can try some things. Do we want to work off a consistent intersection output? Do you have one handy that is small enough to share?

nictru commented:
I sent you a minimal example based on the test configuration via slack. Currently the same bash script is applied to each batch (step 3 of the process):

> 1. Union of all circRNAs
> 2. Split into multiple files (by chromosome or line number)
> 3. Annotate
> 4. Union of annotations
> 5. For each sample-tool combination extract the fitting annotations

I still believe this could be a working approach:

> I tried using bedtools intersect directly on all circRNAs and the GTF file, which was super fast in comparison. The only problem here is, that we need to do some aggregation of all the hits for each circRNA afterwards. I can imagine that this could work well using pandas.groupBy.

Note, that the `bedtools intersect` gives one line for each hit of a given circRNA in the GTF file, so potentially multiple for each circRNA. We could then use some grouping mechanism to perform the main logic of the annotation script (determination of circRNA type and extraction of additional properties).

nictru committed:
Prettier

nictru committed:
Implement bedtools-pandas annotation

nictru committed:
Add transcript ID to annotation

nictru commented:
#95 fixes this better
